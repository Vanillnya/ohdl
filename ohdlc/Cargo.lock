# This file is automatically @generated by Cargo.
# It is not intended for manual editing.
version = 3

[[package]]
name = "ahash"
version = "0.8.11"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "e89da841a80418a9b391ebaea17f5c112ffaaa96f621d2c285b5174da76b9011"
dependencies = [
 "cfg-if",
 "once_cell",
 "version_check",
 "zerocopy",
]

[[package]]
name = "ariadne"
version = "0.4.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "44055e597c674aef7cb903b2b9f6e4cba1277ed0d2d61dae7cd52d7ffa81f8e2"
dependencies = [
 "concolor",
 "unicode-width",
 "yansi",
]

[[package]]
name = "autocfg"
version = "1.2.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "f1fdabc7756949593fe60f30ec81974b613357de856987752631dea1e3394c80"

[[package]]
name = "beef"
version = "0.5.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "3a8241f3ebb85c056b509d4327ad0358fbbba6ffb340bf388f26350aeda225b1"

[[package]]
name = "bitflags"
version = "1.3.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "bef38d45163c2f1dde094a7dfd33ccf595c92905c8f8f4fdc18d06fb1037718a"

[[package]]
name = "bitflags"
version = "2.5.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "cf4b9d6a944f767f8e5e0db018570623c85f3d925ac718db4e06d0187adb21c1"

[[package]]
name = "bumpalo"
version = "3.16.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "79296716171880943b8470b5f8d03aa55eb2e645a4874bdbb28adb49162e012c"

[[package]]
name = "cfg-if"
version = "1.0.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "baf1de4339761588bc0619e3cbc0120ee582ebb74b53b4efbf79117bd2da40fd"

[[package]]
name = "concolor"
version = "0.1.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "0b946244a988c390a94667ae0e3958411fa40cc46ea496a929b263d883f5f9c3"
dependencies = [
 "bitflags 1.3.2",
 "concolor-query",
 "is-terminal",
]

[[package]]
name = "concolor-query"
version = "0.3.3"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "88d11d52c3d7ca2e6d0040212be9e4dbbcd78b6447f535b6b561f449427944cf"
dependencies = [
 "windows-sys 0.45.0",
]

[[package]]
name = "dashmap"
version = "5.5.3"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "978747c1d849a7d2ee5e8adc0159961c48fb7e5db2f06af6723b80123bb53856"
dependencies = [
 "cfg-if",
 "hashbrown 0.14.5",
 "lock_api",
 "once_cell",
 "parking_lot_core",
]

[[package]]
name = "deref-derive"
version = "0.1.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "4255bb7dd538590188bd0aea52e48bd699b19bd90b0d069ec2ced8461fe23273"
dependencies = [
 "proc-macro2",
 "quote",
 "syn 1.0.109",
]

[[package]]
name = "fnv"
version = "1.0.7"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "3f9eec918d3f24069decb9af1554cad7c880e2da24a9afd88aca000531ab82c1"

[[package]]
name = "hashbrown"
version = "0.13.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "43a3c133739dddd0d2990f9a4bdf8eb4b21ef50e4851ca85ab661199821d510e"
dependencies = [
 "ahash",
]

[[package]]
name = "hashbrown"
version = "0.14.5"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "e5274423e17b7c9fc20b6e7e208532f9b19825d82dfd615708b70edd83df41f1"

[[package]]
name = "hermit-abi"
version = "0.3.9"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "d231dfb89cfffdbc30e7fc41579ed6066ad03abda9e567ccafae602b97ec5024"

[[package]]
name = "is-terminal"
version = "0.4.12"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "f23ff5ef2b80d608d61efee834934d862cd92461afc0560dedf493e4c033738b"
dependencies = [
 "hermit-abi",
 "libc",
 "windows-sys 0.52.0",
]

[[package]]
name = "lasso"
version = "0.7.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "4644821e1c3d7a560fe13d842d13f587c07348a1a05d3a797152d41c90c56df2"
dependencies = [
 "dashmap",
 "hashbrown 0.13.2",
]

[[package]]
name = "lazy_static"
version = "1.4.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "e2abad23fbc42b3700f2f279844dc832adb2b2eb069b2df918f455c4e18cc646"

[[package]]
name = "libc"
version = "0.2.154"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "ae743338b92ff9146ce83992f766a31066a91a8c84a45e0e9f21e7cf6de6d346"

[[package]]
name = "lock_api"
version = "0.4.12"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "07af8b9cdd281b7915f413fa73f29ebd5d55d0d3f0155584dade1ff18cea1b17"
dependencies = [
 "autocfg",
 "scopeguard",
]

[[package]]
name = "logos"
version = "0.14.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "161971eb88a0da7ae0c333e1063467c5b5727e7fb6b710b8db4814eade3a42e8"
dependencies = [
 "logos-derive",
]

[[package]]
name = "logos-codegen"
version = "0.14.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "8e31badd9de5131fdf4921f6473d457e3dd85b11b7f091ceb50e4df7c3eeb12a"
dependencies = [
 "beef",
 "fnv",
 "lazy_static",
 "proc-macro2",
 "quote",
 "regex-syntax",
 "syn 2.0.60",
]

[[package]]
name = "logos-derive"
version = "0.14.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "1c2a69b3eb68d5bd595107c9ee58d7e07fe2bb5e360cc85b0f084dedac80de0a"
dependencies = [
 "logos-codegen",
]

[[package]]
name = "nonmax"
version = "0.5.5"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "610a5acd306ec67f907abe5567859a3c693fb9886eb1f012ab8f2a47bef3db51"

[[package]]
name = "ohdlc"
version = "0.1.0"
dependencies = [
 "ariadne",
 "bumpalo",
 "deref-derive",
 "lasso",
 "logos",
 "once_cell",
 "surotto",
]

[[package]]
name = "once_cell"
version = "1.19.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "3fdb12b2476b595f9358c5161aa467c2438859caa136dec86c26fdd2efe17b92"

[[package]]
name = "parking_lot_core"
version = "0.9.10"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "1e401f977ab385c9e4e3ab30627d6f26d00e2c73eef317493c4ec6d468726cf8"
dependencies = [
 "cfg-if",
 "libc",
 "redox_syscall",
 "smallvec",
 "windows-targets 0.52.5",
]

[[package]]
name = "proc-macro2"
version = "1.0.81"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "3d1597b0c024618f09a9c3b8655b7e430397a36d23fdafec26d6965e9eec3eba"
dependencies = [
 "unicode-ident",
]

[[package]]
name = "quote"
version = "1.0.36"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "0fa76aaf39101c457836aec0ce2316dbdc3ab723cdda1c6bd4e6ad4208acaca7"
dependencies = [
 "proc-macro2",
]

[[package]]
name = "redox_syscall"
version = "0.5.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "469052894dcb553421e483e4209ee581a45100d31b4018de03e5a7ad86374a7e"
dependencies = [
 "bitflags 2.5.0",
]

[[package]]
name = "regex-syntax"
version = "0.8.3"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "adad44e29e4c806119491a7f06f03de4d1af22c3a680dd47f1e6e179439d1f56"

[[package]]
name = "scopeguard"
version = "1.2.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "94143f37725109f92c262ed2cf5e59bce7498c01bcc1502d7b9afe439a4e9f49"

[[package]]
name = "smallvec"
version = "1.13.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "3c5e1a9a646d36c3599cd173a41282daf47c44583ad367b8e6837255952e5c67"

[[package]]
name = "surotto"
version = "0.1.0"
source = "git+https://github.com/DasLixou/surotto#51a5c5a71ee44fd3ab2e2a7e71686d9e7534c96c"
dependencies = [
 "nonmax",
]

[[package]]
name = "syn"
version = "1.0.109"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "72b64191b275b66ffe2469e8af2c1cfe3bafa67b529ead792a6d0160888b4237"
dependencies = [
 "proc-macro2",
 "quote",
 "unicode-ident",
]

[[package]]
name = "syn"
version = "2.0.60"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "909518bc7b1c9b779f1bbf07f2929d35af9f0f37e47c6e9ef7f9dddc1e1821f3"
dependencies = [
 "proc-macro2",
 "quote",
 "unicode-ident",
]

[[package]]
name = "unicode-ident"
version = "1.0.12"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "3354b9ac3fae1ff6755cb6db53683adb661634f67557942dea4facebec0fee4b"

[[package]]
name = "unicode-width"
version = "0.1.12"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "68f5e5f3158ecfd4b8ff6fe086db7c8467a2dfdac97fe420f2b7c4aa97af66d6"

[[package]]
name = "version_check"
version = "0.9.4"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "49874b5167b65d7193b8aba1567f5c7d93d001cafc34600cee003eda787e483f"

[[package]]
name = "windows-sys"
version = "0.45.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "75283be5efb2831d37ea142365f009c02ec203cd29a3ebecbc093d52315b66d0"
dependencies = [
 "windows-targets 0.42.2",
]

[[package]]
name = "windows-sys"
version = "0.52.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "282be5f36a8ce781fad8c8ae18fa3f9beff57ec1b52cb3de0789201425d9a33d"
dependencies = [
 "windows-targets 0.52.5",
]

[[package]]
name = "windows-targets"
version = "0.42.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "8e5180c00cd44c9b1c88adb3693291f1cd93605ded80c250a75d472756b4d071"
dependencies = [
 "windows_aarch64_gnullvm 0.42.2",
 "windows_aarch64_msvc 0.42.2",
 "windows_i686_gnu 0.42.2",
 "windows_i686_msvc 0.42.2",
 "windows_x86_64_gnu 0.42.2",
 "windows_x86_64_gnullvm 0.42.2",
 "windows_x86_64_msvc 0.42.2",
]

[[package]]
name = "windows-targets"
version = "0.52.5"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "6f0713a46559409d202e70e28227288446bf7841d3211583a4b53e3f6d96e7eb"
dependencies = [
 "windows_aarch64_gnullvm 0.52.5",
 "windows_aarch64_msvc 0.52.5",
 "windows_i686_gnu 0.52.5",
 "windows_i686_gnullvm",
 "windows_i686_msvc 0.52.5",
 "windows_x86_64_gnu 0.52.5",
 "windows_x86_64_gnullvm 0.52.5",
 "windows_x86_64_msvc 0.52.5",
]

[[package]]
name = "windows_aarch64_gnullvm"
version = "0.42.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "597a5118570b68bc08d8d59125332c54f1ba9d9adeedeef5b99b02ba2b0698f8"

[[package]]
name = "windows_aarch64_gnullvm"
version = "0.52.5"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "7088eed71e8b8dda258ecc8bac5fb1153c5cffaf2578fc8ff5d61e23578d3263"

[[package]]
name = "windows_aarch64_msvc"
version = "0.42.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "e08e8864a60f06ef0d0ff4ba04124db8b0fb3be5776a5cd47641e942e58c4d43"

[[package]]
name = "windows_aarch64_msvc"
version = "0.52.5"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "9985fd1504e250c615ca5f281c3f7a6da76213ebd5ccc9561496568a2752afb6"

[[package]]
name = "windows_i686_gnu"
version = "0.42.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "c61d927d8da41da96a81f029489353e68739737d3beca43145c8afec9a31a84f"

[[package]]
name = "windows_i686_gnu"
version = "0.52.5"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "88ba073cf16d5372720ec942a8ccbf61626074c6d4dd2e745299726ce8b89670"

[[package]]
name = "windows_i686_gnullvm"
version = "0.52.5"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "87f4261229030a858f36b459e748ae97545d6f1ec60e5e0d6a3d32e0dc232ee9"

[[package]]
name = "windows_i686_msvc"
version = "0.42.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "44d840b6ec649f480a41c8d80f9c65108b92d89345dd94027bfe06ac444d1060"

[[package]]
name = "windows_i686_msvc"
version = "0.52.5"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "db3c2bf3d13d5b658be73463284eaf12830ac9a26a90c717b7f771dfe97487bf"

[[package]]
name = "windows_x86_64_gnu"
version = "0.42.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "8de912b8b8feb55c064867cf047dda097f92d51efad5b491dfb98f6bbb70cb36"

[[package]]
name = "windows_x86_64_gnu"
version = "0.52.5"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "4e4246f76bdeff09eb48875a0fd3e2af6aada79d409d33011886d3e1581517d9"

[[package]]
name = "windows_x86_64_gnullvm"
version = "0.42.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "26d41b46a36d453748aedef1486d5c7a85db22e56aff34643984ea85514e94a3"

[[package]]
name = "windows_x86_64_gnullvm"
version = "0.52.5"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "852298e482cd67c356ddd9570386e2862b5673c85bd5f88df9ab6802b334c596"

[[package]]
name = "windows_x86_64_msvc"
version = "0.42.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "9aec5da331524158c6d1a4ac0ab1541149c0b9505fde06423b02f5ef0106b9f0"

[[package]]
name = "windows_x86_64_msvc"
version = "0.52.5"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "bec47e5bfd1bff0eeaf6d8b485cc1074891a197ab4225d504cb7a1ab88b02bf0"

[[package]]
name = "yansi"
version = "1.0.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "cfe53a6657fd280eaa890a3bc59152892ffa3e30101319d168b781ed6529b049"

[[package]]
name = "zerocopy"
version = "0.7.32"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "74d4d3961e53fa4c9a25a8637fc2bfaf2595b3d3ae34875568a5cf64787716be"
dependencies = [
 "zerocopy-derive",
]

[[package]]
name = "zerocopy-derive"
version = "0.7.32"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "9ce1b18ccd8e73a9321186f97e46f9f04b778851177567b1975109d26a08d2a6"
dependencies = [
 "proc-macro2",
 "quote",
 "syn 2.0.60",
]
