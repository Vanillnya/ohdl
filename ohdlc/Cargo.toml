[package]
name = "ohdlc"
version = "0.1.0"
edition = "2021"

[dependencies]
ariadne = { version = "0.4.0", features = ["auto-color"] }
bumpalo = { version = "3.14.0", features = ["collections"] }
deref-derive = "0.1.0"
lasso = { version = "0.7.2", features = ["multi-threaded"] }
logos = "0.14.0"
once_cell = "1.19.0"
surotto = { git = "https://github.com/DasLixou/surotto" }
