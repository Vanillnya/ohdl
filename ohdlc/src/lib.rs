#![deny(rust_2018_idioms)]

use message::Messages;

// TODO: can we get unused errors back? :3
pub mod ast;
pub mod ir;
pub mod lexer;
pub mod message;
pub mod parser;
pub mod span;
pub mod symbol;

pub static MESSAGES: Messages = Messages::new();

#[derive(Clone)]
pub struct Source<'s>(pub String, pub &'s str);
