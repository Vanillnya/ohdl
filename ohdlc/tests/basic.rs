use ariadne::{Label, Report};
use bumpalo::Bump;
use insta::assert_debug_snapshot;
use ohdlc::{
    ir::{
        import_bucket::ImportBucket,
        name_lookup::NameLookup,
        registries::{ModuleRegistry, RoughEntityRegistry, RoughTypeRegistry},
        stages::{flatten_lookup::FlattenLookupStage, refine::RefineStage, rough::RoughStage},
    },
    lexer::Lexer,
    parser::Parser,
    Source, MESSAGES,
};

#[test]
fn main() {
    let source = Source("work.ohd".to_owned(), include_str!("basic.ohd"));

    println!("[STAGE] Lexer");

    let lexer = Lexer::new(&source.1);
    report_messages(&source);
    let lexer = lexer.unwrap();

    println!("[STAGE] Parser");

    let parser_arena = Bump::new();

    let mut parser = Parser::new(&parser_arena, source.clone(), lexer);

    let root = parser.parse();
    let root = match root {
        Ok(tree) => tree,
        Err(messages) => {
            MESSAGES.extend(messages);
            report_messages(&source);
            panic!();
        }
    };

    assert_debug_snapshot!(root);

    let ir_arena = Bump::new();
    let mut module_reg = ModuleRegistry::default();
    let mut type_reg = RoughTypeRegistry::default();
    let mut entity_reg = RoughEntityRegistry::default();
    let mut name_lookup = NameLookup::new();
    let mut import_bucket = ImportBucket::new();

    {
        let rough = RoughStage {
            arena: &ir_arena,
            name_lookup: &mut name_lookup,
            import_bucket: &mut import_bucket,
            module_reg: &mut module_reg,
            type_reg: &mut type_reg,
            entity_reg: &mut entity_reg,
            root: &root,
        };
        rough.lower();
        report_messages(&source);
    }

    let name_lookup = {
        let resolve = FlattenLookupStage {
            module_reg: &module_reg,
            name_lookup,
            import_bucket,
            resolvables: Vec::new(),
        };
        let lookup = resolve.lower();
        report_messages(&source);
        lookup
    };
    let name_lookup = name_lookup.unwrap();

    let (type_reg, entity_reg) = {
        let refine = RefineStage {
            arena: &ir_arena,
            name_lookup: &name_lookup,
            module_registry: &module_reg,
        };

        let type_reg = refine.refine_types(type_reg);
        report_messages(&source);

        let entity_reg = refine.refine_entities(entity_reg);
        report_messages(&source);

        (type_reg, entity_reg)
    };

    assert_debug_snapshot!(type_reg);
    assert_debug_snapshot!(entity_reg);
}

fn report_messages(source: &Source<'_>) {
    MESSAGES.drain(|msg| {
        let filename = source.0.as_str();

        let report =
            Report::build(msg.kind, filename, msg.location.0)
                .with_message(msg.message)
                .with_labels(msg.labels.into_iter().map(|label| {
                    Label::new((filename, label.span.into())).with_message(label.message)
                }))
                .finish();

        report
            .eprint((filename, ariadne::Source::from(source.1)))
            .unwrap();
    });
}
