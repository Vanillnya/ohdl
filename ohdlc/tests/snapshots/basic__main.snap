---
source: tests/basic.rs
expression: root
---
[
    Module(
        Module {
            name: `std`:4..7,
            items: [
                Enum(
                    Enum {
                        name: `Logical`:20..27,
                        variants: [
                            `High`:39..43,
                            `Low`:54..57,
                        ],
                    },
                ):15..68,
            ],
        },
    ):0..73,
    Use(
        Use {
            path: Path(
                [
                    PathSegment(
                        `std`:76..79,
                    ),
                    PathSegment(
                        `Logical`:81..88,
                    ),
                ],
                Root,
            ):74..89,
        },
    ):70..96,
    Module(
        Module {
            name: `x`:97..98,
            items: [
                Use(
                    Use {
                        path: Path(
                            [
                                PathSegment(
                                    `y`:110..111,
                                ),
                            ],
                            Local,
                        ):110..112,
                    },
                ):106..115,
            ],
        },
    ):93..122,
    Module(
        Module {
            name: `y`:123..124,
            items: [
                Use(
                    Use {
                        path: Path(
                            [
                                PathSegment(
                                    `x`:136..137,
                                ),
                            ],
                            Local,
                        ):136..138,
                    },
                ):132..141,
            ],
        },
    ):119..151,
    Record(
        Record {
            name: `MyRecord`:152..160,
            fields: [
                Field {
                    name: `a`:168..169,
                    ty: Type {
                        path: Path(
                            [
                                PathSegment(
                                    `Logical`:171..178,
                                ),
                            ],
                            Local,
                        ):171..179,
                    }:171..179,
                }:168..179,
                Field {
                    name: `b`:185..186,
                    ty: Type {
                        path: Path(
                            [
                                PathSegment(
                                    `Logical`:188..195,
                                ),
                            ],
                            Local,
                        ):188..196,
                    }:188..196,
                }:185..196,
            ],
        },
    ):145..207,
    Enum(
        Enum {
            name: `MyEnum`:208..214,
            variants: [
                `A`:222..223,
                `B`:230..231,
                `C`:238..239,
                `D`:246..247,
                `E`:254..255,
            ],
        },
    ):203..269,
    Entity(
        Entity {
            name: `HalfAdder`:270..279,
            ports: [
                Port {
                    kind: Input:287..289,
                    name: `a`:291..292,
                    ty: Type {
                        path: Path(
                            [
                                PathSegment(
                                    `Logical`:294..301,
                                ),
                            ],
                            Local,
                        ):294..302,
                    }:294..302,
                }:287..302,
                Port {
                    kind: Input:308..310,
                    name: `b`:312..313,
                    ty: Type {
                        path: Path(
                            [
                                PathSegment(
                                    `Logical`:315..322,
                                ),
                            ],
                            Local,
                        ):315..323,
                    }:315..323,
                }:308..323,
                Port {
                    kind: Output:329..332,
                    name: `o`:333..334,
                    ty: Type {
                        path: Path(
                            [
                                PathSegment(
                                    `Logical`:336..343,
                                ),
                            ],
                            Local,
                        ):336..344,
                    }:336..344,
                }:329..344,
                Port {
                    kind: Output:350..353,
                    name: `c`:354..355,
                    ty: Type {
                        path: Path(
                            [
                                PathSegment(
                                    `Logical`:357..364,
                                ),
                            ],
                            Local,
                        ):357..365,
                    }:357..365,
                }:350..365,
            ],
        },
    ):263..376,
    Arch(
        Arch {
            name: `RTL`:377..380,
            ty: Type {
                path: Path(
                    [
                        PathSegment(
                            `HalfAdder`:385..394,
                        ),
                    ],
                    Local,
                ):385..396,
            }:385..396,
            stmts: [
                Assign(
                    AssignStmt {
                        assignee: `o`:402..403,
                        value: Binary {
                            left: Primary(
                                `a`:407..408,
                            ),
                            right: Primary(
                                `b`:413..414,
                            ),
                            operator: Xor,
                        },
                    },
                ):402..422,
                Assign(
                    AssignStmt {
                        assignee: `c`:421..422,
                        value: Binary {
                            left: Primary(
                                `a`:426..427,
                            ),
                            right: Primary(
                                `b`:432..433,
                            ),
                            operator: And,
                        },
                    },
                ):421..437,
            ],
        },
    ):372..444,
    Use(
        Use {
            path: Path(
                [
                    PathSegment(
                        `a`:445..446,
                    ),
                    PathSegment(
                        `b`:448..449,
                    ),
                    PathSegment(
                        `Lol`:451..454,
                    ),
                ],
                Local,
            ):445..455,
        },
    ):441..462,
    Module(
        Module {
            name: `a`:463..464,
            items: [
                Use(
                    Use {
                        path: Path(
                            [
                                PathSegment(
                                    `other`:476..481,
                                ),
                                PathSegment(
                                    `b`:483..484,
                                ),
                            ],
                            Local,
                        ):476..485,
                    },
                ):472..488,
            ],
        },
    ):459..495,
    Module(
        Module {
            name: `other`:496..501,
            items: [
                Module(
                    Module {
                        name: `b`:513..514,
                        items: [
                            Entity(
                                Entity {
                                    name: `Lol`:533..536,
                                    ports: [],
                                },
                            ):526..547,
                        ],
                    },
                ):509..550,
            ],
        },
    ):492..560,
    Entity(
        Entity {
            name: `FullAdder`:561..570,
            ports: [
                Port {
                    kind: Input:578..580,
                    name: `a`:582..583,
                    ty: Type {
                        path: Path(
                            [
                                PathSegment(
                                    `Logical`:585..592,
                                ),
                            ],
                            Local,
                        ):585..593,
                    }:585..593,
                }:578..593,
                Port {
                    kind: Input:599..601,
                    name: `b`:603..604,
                    ty: Type {
                        path: Path(
                            [
                                PathSegment(
                                    `Logical`:606..613,
                                ),
                            ],
                            Local,
                        ):606..614,
                    }:606..614,
                }:599..614,
                Port {
                    kind: Input:620..622,
                    name: `ci`:624..626,
                    ty: Type {
                        path: Path(
                            [
                                PathSegment(
                                    `Logical`:628..635,
                                ),
                            ],
                            Local,
                        ):628..636,
                    }:628..636,
                }:620..636,
                Port {
                    kind: Output:642..645,
                    name: `o`:646..647,
                    ty: Type {
                        path: Path(
                            [
                                PathSegment(
                                    `Logical`:649..656,
                                ),
                            ],
                            Local,
                        ):649..657,
                    }:649..657,
                }:642..657,
                Port {
                    kind: Output:663..666,
                    name: `co`:667..669,
                    ty: Type {
                        path: Path(
                            [
                                PathSegment(
                                    `Logical`:671..678,
                                ),
                            ],
                            Local,
                        ):671..679,
                    }:671..679,
                }:663..679,
            ],
        },
    ):554..690,
    Arch(
        Arch {
            name: `RTL`:691..694,
            ty: Type {
                path: Path(
                    [
                        PathSegment(
                            `FullAdder`:699..708,
                        ),
                    ],
                    Local,
                ):699..710,
            }:699..710,
            stmts: [
                Place(
                    PlaceStmt {
                        entity_ty: Type {
                            path: Path(
                                [
                                    PathSegment(
                                        `HalfAdder`:716..725,
                                    ),
                                ],
                                Local,
                            ):716..726,
                        }:716..726,
                        arch_ty: Type {
                            path: Path(
                                [
                                    PathSegment(
                                        `RTL`:726..729,
                                    ),
                                ],
                                Local,
                            ):726..730,
                        }:726..730,
                        links: [
                            PlaceLink {
                                src: `a`:742..743,
                                arrow_span: Span(
                                    744,
                                    746,
                                ),
                                link: Ingoing(
                                    Primary(
                                        `a`:747..748,
                                    ):747..749,
                                ),
                            }:742..749,
                            PlaceLink {
                                src: `b`:759..760,
                                arrow_span: Span(
                                    761,
                                    763,
                                ),
                                link: Ingoing(
                                    Primary(
                                        `b`:764..765,
                                    ):764..766,
                                ),
                            }:759..766,
                            PlaceLink {
                                src: `o`:776..777,
                                arrow_span: Span(
                                    778,
                                    780,
                                ),
                                link: Outgoing(
                                    NewSignal(
                                        `s`:788..789,
                                    ):781..790,
                                ),
                            }:776..790,
                            PlaceLink {
                                src: `c`:800..801,
                                arrow_span: Span(
                                    802,
                                    804,
                                ),
                                link: Outgoing(
                                    NewSignal(
                                        `cb`:812..814,
                                    ):805..815,
                                ),
                            }:800..815,
                        ],
                    },
                ):716..839,
                Place(
                    PlaceStmt {
                        entity_ty: Type {
                            path: Path(
                                [
                                    PathSegment(
                                        `HalfAdder`:830..839,
                                    ),
                                ],
                                Local,
                            ):830..840,
                        }:830..840,
                        arch_ty: Type {
                            path: Path(
                                [
                                    PathSegment(
                                        `RTL`:840..843,
                                    ),
                                ],
                                Local,
                            ):840..844,
                        }:840..844,
                        links: [
                            PlaceLink {
                                src: `a`:856..857,
                                arrow_span: Span(
                                    858,
                                    860,
                                ),
                                link: Ingoing(
                                    Primary(
                                        `s`:861..862,
                                    ):861..863,
                                ),
                            }:856..863,
                            PlaceLink {
                                src: `b`:873..874,
                                arrow_span: Span(
                                    875,
                                    877,
                                ),
                                link: Ingoing(
                                    Primary(
                                        `ci`:878..880,
                                    ):878..881,
                                ),
                            }:873..881,
                            PlaceLink {
                                src: `o`:891..892,
                                arrow_span: Span(
                                    893,
                                    895,
                                ),
                                link: Outgoing(
                                    Ref(
                                        `o`:896..897,
                                    ):896..898,
                                ),
                            }:891..898,
                            PlaceLink {
                                src: `c`:908..909,
                                arrow_span: Span(
                                    910,
                                    912,
                                ),
                                link: Outgoing(
                                    NewSignal(
                                        `cc`:920..922,
                                    ):913..923,
                                ),
                            }:908..923,
                        ],
                    },
                ):830..940,
                Assign(
                    AssignStmt {
                        assignee: `co`:938..940,
                        value: Binary {
                            left: Primary(
                                `cb`:944..946,
                            ),
                            right: Primary(
                                `cc`:950..952,
                            ),
                            operator: Or,
                        },
                    },
                ):938..956,
            ],
        },
    ):686..956,
]
